src/common/dir.c

```diff
@@ -4,143 +4,189 @@
 #include "common/dir.h"
+#include "common/buf.h"
+#include "common/list.h"
+#include "common/mem.h"
+#include "common/string-helpers.h"
+#include "labwc.h"
 
 struct dir {
 	const char *prefix;
+	const char *default_prefix;
 	const char *path;
 };
 
 static struct dir config_dirs[] = {
-	{ "XDG_CONFIG_HOME", "labwc" },
-	{ "HOME", ".config/labwc" },
-	{ "XDG_CONFIG_DIRS", "labwc" },
-	{ NULL, "/etc/xdg/labwc" },
-	{ NULL, NULL }
+	{
+		.prefix = "XDG_CONFIG_HOME",
+		.default_prefix = "$HOME/.config",
+		.path = "labwc"
+	}, {
+		.prefix = "XDG_CONFIG_DIRS",
+		.default_prefix = "/etc/xdg",
+		.path = "labwc",
+	}, {
+		.path = NULL,
+	}
 };
 
 static struct dir theme_dirs[] = {
-	{ "XDG_DATA_HOME", "themes" },
-	{ "HOME", ".local/share/themes" },
-	{ "HOME", ".themes" },
-	{ "XDG_DATA_DIRS", "themes" },
-	{ NULL, "/usr/share/themes" },
-	{ NULL, "/usr/local/share/themes" },
-	{ NULL, "/opt/share/themes" },
-	{ NULL, NULL }
+	{
+		.prefix = "XDG_DATA_HOME",
+		.default_prefix = "$HOME/.local/share",
+		.path = "themes",
+	}, {
+		.prefix = "HOME",
+		.path = ".themes",
+	}, {
+		.prefix = "XDG_DATA_DIRS",
+		.default_prefix = "/usr/share:/usr/local/share:/opt/share",
+		.path = "themes",
+	}, {
+		.path = NULL,
+	}
 };
 
-static bool
-isdir(const char *path)
-{
-	struct stat st;
-	return (!stat(path, &st) && S_ISDIR(st.st_mode));
-}
-
 struct ctx {
 	void (*build_path_fn)(struct ctx *ctx, char *prefix, const char *path);
+	const char *filename;
 	char *buf;
 	size_t len;
 	struct dir *dirs;
 	const char *theme_name;
+	struct wl_list *list;
 };
 
+struct wl_list *paths_get_prev(struct wl_list *elm) { return elm->prev; }
+struct wl_list *paths_get_next(struct wl_list *elm) { return elm->next; }
+
 static void
 build_config_path(struct ctx *ctx, char *prefix, const char *path)
 {
-	if (!prefix) {
-		snprintf(ctx->buf, ctx->len, "%s", path);
-	} else {
-		snprintf(ctx->buf, ctx->len, "%s/%s", prefix, path);
-	}
+	assert(prefix);
+	snprintf(ctx->buf, ctx->len, "%s/%s/%s", prefix, path, ctx->filename);
 }
 
 static void
 build_theme_path(struct ctx *ctx, char *prefix, const char *path)
 {
-	if (!prefix) {
-		snprintf(ctx->buf, ctx->len, "%s/%s/openbox-3", path,
-			ctx->theme_name);
-	} else {
-		snprintf(ctx->buf, ctx->len, "%s/%s/%s/openbox-3", prefix, path,
-			ctx->theme_name);
-	}
+	assert(prefix);
+	snprintf(ctx->buf, ctx->len, "%s/%s/%s/openbox-3/%s", prefix, path,
+		ctx->theme_name, ctx->filename);
 }
 
-static char *
+static void
 find_dir(struct ctx *ctx)
 {
 	char *debug = getenv("LABWC_DEBUG_DIR_CONFIG_AND_THEME");
 
 	for (int i = 0; ctx->dirs[i].path; i++) {
 		struct dir d = ctx->dirs[i];
-		if (!d.prefix) {
-			/* handle /etc/xdg... */
-			ctx->build_path_fn(ctx, NULL, d.path);
+		struct buf prefix;
+		buf_init(&prefix);
+
+		/*
+		 * Replace (rather than augment) $HOME/.config with
+		 * $XDG_CONFIG_HOME if defined, and so on for the other
+		 * XDG Base Directories.
+		 */
+		buf_add(&prefix, getenv(d.prefix) ? : d.default_prefix);
```

Comment:
`?:` without a second expression seems to be a GNU extension.
```suggestion
		char *pfxenv = getenv(d.prefix);
		buf_add(&prefix, pfxenv ? pfxenv : d.default_prefix);
```

Reply:
Done.
Yes, should avoid that really.... have just become rather fond of it.